File: src/FASTCAMPUS/ALGO/D25/P15663.js
const fs = require('fs');
class Scanner {
  constructor(fileName){
    this.read = fs.readFileSync(fileName).toString().split('\n');
    this.stringToken = [];
  }
  next(){
    if(this.stringToken.length === 0){
      if(this.read.length > 0){
        this.stringToken = this.read.shift().split(" ");
      } else {
        return null;
      }
    } 
    return this.stringToken.shift();
  }

  nextNumber(){
    const next = this.next();
    if(next){
      return Number(next);
    } else {
      return null;
    }
    
  }

  nextLine(){
    if(this.read.length > 0){
      return this.read.shift();
    } else {
      return null;
    }
    
  }
}
const scanner = new Scanner('stdin.txt');

let N, M, Arr, visited;
const input = function (){
  N = scanner.nextNumber();
  M = scanner.nextNumber();
  Arr = [];
  visited = [];
  for(let i = 0; i < N; i ++){
    Arr[i] = scanner.nextNumber();
  }
}
const dfs = function(arr){
  if(arr.length == M){
    console.log(arr.join(" "));
    return;
  }
  let last = 0;
  for(let i = 0; i < N; i ++){
    if(visited.indexOf(i) != -1) continue;
    if(last === Arr[i]) continue;
    arr.push(Arr[i]);
    visited.push(i);
    last = Arr[i];
    dfs(arr);
    arr.pop();
    visited.pop();
  }
};
const pro = function (){
  Arr.sort((a,b) => {
    return a - b;
  })
  dfs([]);
}

const solve = function(){
  input();
  pro();
}

solve();

File: src/FASTCAMPUS/ALGO/D25/P15663.java
package FASTCAMPUS.ALGO.D25;
import java.io.*;
import java.util.*;

public class P15663 {
    static int N;
    static int M;
    static int A[];
    static ArrayList<String> arr;
    static ArrayList<Integer> in;
    static boolean [] visited;
    static void input() {
        FastReader scan = new FastReader();
        N = scan.nextInt();
        M = scan.nextInt();
        A = new int[N];
        arr = new ArrayList<String>();
        in = new ArrayList<Integer>();
        visited = new boolean [N];
        for(int i = 0; i < N; i ++){
            A[i] = scan.nextInt();
        }
    }

    public static void dfs(){
        if(in.size() == M){
            StringBuilder sb = new StringBuilder();
            for (int i : in) {
                sb.append(i + " ");
            }

            System.out.println(
                sb
            );
            return;
        }
        int last_cand = 0;
        for(int i = 0; i < N ; i ++){
            if (A[i] == last_cand) continue;
            if(visited[i] == true) continue;
            visited[i] = true;
            in.add(A[i]);
            last_cand = A[i];
            dfs();
            in.remove(in.size() - 1);
            visited[i] = false;
        }
    }
    public static void pro(){
        ArrayList<Integer> in = new ArrayList<Integer>();
        Arrays.sort(A);
        dfs();
        // for (String answer : arr) {
        //     System.out.println(answer.trim());
        // }
    }
    public static void main(String[] args) {
        input();
        pro();
    }
    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}